File: src/content/docs/topics/cs/coding-problems/graphs/547-number-of-provinces-approach2.ts
function assert(condition: boolean, msg: string = ''): void {
    if (!condition) throw new Error(msg || 'Assertion failed');
}

function findCircleNum(isConnected: number[][]): number {
    const n = isConnected.length;
    const visited = new Array(n).fill(false);
    let provinces = 0;

    function dfs(city: number): void {
        visited[city] = true;
        for (let neighbor = 0; neighbor < n; neighbor++) {
            if (isConnected[city][neighbor] === 1 && !visited[neighbor]) {
                dfs(neighbor);
            }
        }
    }

    for (let city = 0; city < n; city++) {
        if (!visited[city]) {
            provinces++;
            dfs(city);
        }
    }
    return provinces;
}

assert(findCircleNum([[1,1,0],[1,1,0],[0,0,1]]) === 2);
assert(findCircleNum([[1,0,0],[0,1,0],[0,0,1]]) === 3);
assert(findCircleNum([[1,1,1],[1,1,1],[1,1,1]]) === 1);
assert(findCircleNum([[1]]) === 1);
assert(findCircleNum([[1,1,0],[1,1,1],[0,1,1]]) === 1);
console.log('all tests pass');

File: src/content/docs/topics/cs/coding-problems/graphs/547-number-of-provinces-approach2.py
def find_circle_num(isConnected: list[list[int]]) -> int:
    n = len(isConnected)
    visited = [False] * n
    provinces = 0

    def dfs(city):
        visited[city] = True
        for neighbor in range(n):
            if isConnected[city][neighbor] == 1 and not visited[neighbor]:
                dfs(neighbor)

    for city in range(n):
        if not visited[city]:
            provinces += 1
            dfs(city)
    return provinces

assert find_circle_num([[1,1,0],[1,1,0],[0,0,1]]) == 2
assert find_circle_num([[1,0,0],[0,1,0],[0,0,1]]) == 3
assert find_circle_num([[1,1,1],[1,1,1],[1,1,1]]) == 1
assert find_circle_num([[1]]) == 1
assert find_circle_num([[1,1,0],[1,1,1],[0,1,1]]) == 1
print("all tests pass")
